### PR TITLE
CIWEMB-69: Adding the Company entity

### PR DESCRIPTION
**Pull request by omarabuhussein**
## Overview

Creating the Company entity to serve as a base for Multi companies accounting. The work in this PR is for all the tickets from CIWEMB-69, CIWEMB-70, CIWEMB-71 and CIWEMB-72 given it make sense to have them done all in one go, where each ticket work is split into its own commit (though I kept same ticket number on all of the commits).

## Before

~~

## After

A new entity called `Company` is created, it contains the following fields:

- `contact_id`: To hold the organization (company) id, which should point to an existing organization contact record .
- `invoice_template_id`: We still not sure how the end solution would look like, but the idea is to be able to have a different invoice template for each Company, so this field will decide which invoice template will be used for each company. (any Message template can be used, thus this field have a foreign key on civicrm_msg_template table).
- `invoice_prefix` and `next_invoice_number`: Since each company will have its own invoicing, these two fields will be used to calculate the next invoice number for each company, they will be used instead of CiviCRM core ones.
- `creditnote_prefix` and `next_creditnote_number`: Same as the fields above, but will be used for credit notes in the future. 

Pages to list, delete and add/update Companies are also created and all of them require "Administer CiviCRM" permission to be accessed.

- Company listing menu item:

image

- Company listing page:
image

- Adding new company:

- Editing existing company:
image

- Delete a company:

image

## Technical Details

The initial idea was to use the Domain entity in CiviCRM, and with the help of `cg_extend_objects` option group we can add Custom fields to the Domain entity to the add the remaining necessary fields, such as `invoice_template_id`,`invoice_prefix` and `next_invoice_number`, and then to use Searchkit and Form builder to create the CRUD pages. 

The main reason was to benefit from what the Domain entity already provides,  first because it is already there so what is the point of creating new entity, and second to benefit from the `tokens`  used in the invoice template such as `domain_organization` and `domain_street_address `.

But there are some issues with all of the above:

1- I did an investigation to use Searchkit and Form builder for another entity not long ago, and my conclusion was:  to make the needed pages and forms to manage any entity, a lot of coding is still needed, and it is a much uglier code than the one we will end up writing if we did things from scratch. The only difference is that using Searchkit will provide the user with a bit of flexibility on changing the table that shows the entity records, and that is not much of a benefit to justify using it. 

2- The tokens (like the ones mentioned above) turned out to be useless, given Civi evaluate them based on the Domain configuration in the civicrm.settings.php file, and if I want to change that then I will have to implement tokenValues hook, which is something I can still do even if new entity was created instead of using the Domain one.

3- Thus it brings me to the point I raised during the specs discussion, where I said it is more neat and cleaner to create new entity and handle everything in code instead of using the Domain one + Custom fields, thus I ended up doing the latter instead of the former in this PR.

### CODE REVIEW

File: multicompanyaccounting.civix.php [hunk 1]
```diff
@@ -449,5 +449,11 @@ function _multicompanyaccounting_civix_civicrm_alterSettingsFolders(&$metaDataFo
  * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_entityTypes
  */
 function _multicompanyaccounting_civix_civicrm_entityTypes(&$entityTypes) {
-  $entityTypes = array_merge($entityTypes, []);
+  $entityTypes = array_merge($entityTypes, [
+    'CRM_Multicompanyaccounting_DAO_Company' => [
+      'name' => 'Company',
+      'class' => 'CRM_Multicompanyaccounting_DAO_Company',
+      'table' => 'civicrm_company',
```
Comment (olayiwola-compucorp) [inline multicompanyaccounting.civix.php]: The table name in other places is `multicompanyaccounting_company`

File: xml/schema/CRM/Multicompanyaccounting/Company.entityType.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+// This file declares a new entity type. For more details, see "hook_civicrm_entityTypes" at:
+// https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_entityTypes
+return [
+  [
+    'name' => 'Company',
+    'class' => 'CRM_Multicompanyaccounting_DAO_Company',
+    'table' => 'civicrm_company',
```
Comment (olayiwola-compucorp) [inline xml/schema/CRM/Multicompanyaccounting/Company.entityType.php]: Same here, the table name in other places is `multicompanyaccounting_company`

File: CRM/Multicompanyaccounting/Page/Company.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<?php
+use CRM_Multicompanyaccounting_ExtensionUtil as E;
```
Comment (olayiwola-compucorp) [inline CRM/Multicompanyaccounting/Page/Company.php]: unused import

**Review by olayiwola-compucorp — CHANGES_REQUESTED** [review]

File: CRM/Multicompanyaccounting/Page/Company.php (hunk 3 above)
Comment (omarabuhussein) [reply CRM/Multicompanyaccounting/Page/Company.php]: this is already removed in one of the commits

File: multicompanyaccounting.civix.php (hunk 1 above)
Comment (omarabuhussein) [reply multicompanyaccounting.civix.php]: fixed

File: xml/schema/CRM/Multicompanyaccounting/Company.entityType.php (hunk 2 above)
Comment (omarabuhussein) [reply xml/schema/CRM/Multicompanyaccounting/Company.entityType.php]: fixed

**Review by olayiwola-compucorp — APPROVED** [review]

**Pull request merged**